### PR TITLE
feat!: add device-class-specific sensor state subtypes

### LINKED ISSUE #717: Add device-class-specific sensor state subtypes
## Description

The codegen now generates a single `SensorState(StringBaseState)` with a `device_class: SensorDeviceClass | None` field and a polymorphic `native_value` type. However, different device classes have predictable value types:

- `temperature`, `humidity`, `energy`, etc. → `float`
- `enum` → `str` (from a known options list)
- `timestamp` → `ZonedDateTime`
- `date` → `Date`

## Motivation

Users writing automations want type-safe access to sensor values without manual casting. A `TemperatureSensorState` with `value: float | None` is more useful than the generic `SensorState` with `value: str | int | float | None | ...`.

## Approach

1. Generate per-device-class subclasses (e.g., `TemperatureSensorState`, `EnumSensorState`) that narrow `value` type and lock `device_class: Literal[SensorDeviceClass.TEMPERATURE]`
2. Update state registry routing to dispatch by `device_class` attribute in the incoming state dict
3. The generic `SensorState` remains as the fallback for unknown device classes

## Context

- `SensorDeviceClass` StrEnum is already generated with all HA device classes
- `device_class` field is typed on `SensorAttributes`
- Value type mapping per device class is documented in HA core: `homeassistant/components/sensor/__init__.py` (`DEVICE_CLASS_STATE_CLASSES` and type annotations)
- Entity wrappers could also specialize (e.g., `TemperatureSensor` entity with narrowed state type)

## Acceptance Criteria

- [ ] Per-device-class state subclasses generated for at least: temperature, humidity, energy, power, battery, enum, timestamp, date
- [ ] `value` field narrowed to the correct type on each subclass
- [ ] State registry routes to specific subclass when `device_class` is present
- [ ] Generic `SensorState` used as fallback for unrecognized device classes
- [ ] Existing code using `SensorState` continues to work (subclasses are `isinstance`-compatible)

### PR DESCRIPTION
**Pull request by NodeJSmith**
## Summary

`SensorState.value` has always been `str | None`, because `SensorState` extends
`StringBaseState` and must accept every sensor domain entity regardless of device class. But
Home Assistant sensors actually carry four distinct value shapes — numeric, enum, timestamp,
and date — so every numeric sensor read forces app authors to cast or work around a type the
tooling can't verify (`state.value + 1` doesn't type-check), and there was no way to ask the
framework for "the sensors whose values are numbers."

This PR adds four narrowed sensor state subclasses reachable by annotation
(`D.StateNew[NumericSensorState]`) and by accessor (`self.states.numeric_sensor`), backed by a
shape classifier ported from Home Assistant's own numeric-value rule so accessor membership and
the framework's runtime behavior agree. See `design/specs/093-sensor-device-class-subtypes/design.md`
for the full design rationale, alternatives considered, and edge-case analysis.

### Notable Changes

- Fixed a flaky caplog test — `enable_basic_logging()` could leave `propagate=False` set on the
  `hassette` logger from an earlier test, causing intermittent log-capture failures unrelated to
  this feature.

### Sensor value shapes

- Added `NumericSensorState`, `EnumSensorState`, `TimestampSensorState`, and `DateSensorState`
  (`src/hassette/models/states/sensor_shapes.py`) — narrowed views of the `sensor` domain, none of
  which register in the state catalog (`resolve(domain="sensor")` still returns `SensorState`).
- Added `classify_sensor_shape`, a single classifier that both accessor membership and DI shape
  validation read from, so an entity's declared type and its runtime type never disagree. Its
  numeric branch is a direct port of Home Assistant's `_numeric_state_expected` predicate rather
  than a device-class allowlist — a naive device-class-only rule would drop ~27% of the real
  numeric population (sensors with no `device_class` but a `state_class` or unit).
- Sensors with device class `uptime` classify as the timestamp shape rather than getting a
  dedicated class — HA reports uptime as a timestamp.
- `StateManager` gained four accessors — `numeric_sensor`, `enum_sensor`, `timestamp_sensor`,
  `date_sensor` — each a `DomainStates` containing exactly the sensors whose metadata *and*
  successful conversion match that shape, preserving the `Mapping` invariant
  `len(m) == len(list(m))` even for a sensor whose metadata claims numeric but whose value fails
  to parse.
- Direct lookup of a non-member entity through a narrowed accessor now raises
  `EntityNotInViewError` (subclasses both `KeyError` and the state-error hierarchy, so
  `Mapping.get()` still returns `None`), naming the entity, its actual device class, and the
  accessor's expected shape. Iteration silently skips non-members at `debug` level.
- Dependency injection now raises when an annotated shape class doesn't match an entity's actual
  value shape, even where coercion would otherwise silently succeed (e.g. annotating a
  temperature sensor as `EnumSensorState` would coerce a float to a string without this check).
  An entity that classifies as "unknown" (no device class, no `state_class`, no unit) doesn't
  raise — there's no shape claim to contradict.
- Codegen now generates `NON_NUMERIC_DEVICE_CLASSES` from Home Assistant core instead of hand
  maintaining it, and snapshots the source text of HA's `_numeric_state_expected`
  (`codegen/snapshots/numeric_state_expected.py.txt`) so a freshness check fails loudly if
  upstream logic changes, forcing a human re-verification of the port at version-bump time.

### Breaking changes

- `SensorAttributes` no longer declares `native_value`, `native_unit_of_measurement`, or
  `suggested_display_precision` — these were always dead duplicates of the base `State.value`/
  `unit_of_measurement` fields (removed via a new per-domain TOML `remove = true` override
  applied to `sensor` only; `native_value` stays live on the five other domains that still use
  it).
- `StateRegistry.register`, `StateRegistry.resolve`, and `register_state_converter` drop their
  `device_class` parameter, and `StateKey` drops its `device_class` field. The dimension was added in passing by #197
  (2025-12-13), populated by nothing (all 55 entries had `device_class=None`), and consumed by
  nothing (every production `resolve()` call site passed `domain` only) — leaving it would have
  been permanently misleading once this design's real device-class logic landed elsewhere.
  `StateKey` itself survives as a one-field frozen dataclass (`domain` only); its export names,
  `isinstance` checks, and role as the yield type of `DomainStates` iteration are unchanged.

See the full footer below for the machine-readable breaking-change summary.

### Housekeeping (post-implementation cleanup)

- Deduplicated codegen AST-parse boilerplate (`_parse_module_or_none`) and
  `RegistryValidationIssue` construction (`_issue()` helper in `validation.py`).
- Removed `StateReader.num_domain_states`/`__contains__` and their `StateProxy`
  implementations — these became orphaned once `DomainStates` stopped delegating to them for
  membership checks (tracked as KI-001 during orchestration, resolved before merge).

### Deferred follow-ups

Three findings from the final clean-code pass were deliberately left for dedicated, reviewed
passes rather than folded into this branch (each documented with full rationale in
`design/specs/093-sensor-device-class-subtypes/known-issues.md`):

- #1545 — the three new sensor-shape exceptions share identical `__init__` boilerplate; a shared
  base class would diverge from this design's documented "one class per error, explicit `__init__`"
  convention without a separate design review.
- #1546 — `run_pipeline` in codegen is a pre-existing ~180-line function mixing six
  responsibilities; this branch only inserted a small block into it, and splitting it is a
  standalone refactor outside spec 093's scope.
- #1547 — the `HA_CORE_PATH`/`sys.path.insert` boilerplate duplicated across 14 codegen test
  files predates this branch, which only touched 2 of the 14; deduplicating just the touched
  files would make them diverge from the rest.
- #1548 — a circular-import workaround between `conversion/validation.py` and
  `models/states/base.py`, discovered during the same clean-code pass, unrelated to the sensor
  shapes feature itself.

## Breaking Changes

BREAKING CHANGE: `SensorAttributes` no longer declares `native_value`, `native_unit_of_measurement`, or `suggested_display_precision` (they duplicated the base `State.value`/`unit_of_measurement` fields and are removed for the `sensor` domain only). `StateRegistry.register` and `register_state_converter` drop their unused `device_class` keyword parameter, and `StateKey` drops its `device_class` field (kept as a one-field frozen dataclass with `domain` only — its export names and `isinstance` checks are unchanged).
#### What changes for callers
- Code reading `sensor_state.attributes.native_value`, `.native_unit_of_measurement`, or `.suggested_display_precision` must switch to `sensor_state.value` and `sensor_state.attributes.unit_of_measurement` — those fields already carried the same data.
- Code calling `StateRegistry.register(..., device_class=...)`, `StateRegistry.resolve(..., device_class=...)`, or `register_state_converter(..., device_class=...)` must drop the `device_class` keyword — it was never populated or read by any production call site.
- Code constructing `StateKey(domain=..., device_class=...)` directly must drop `device_class` — `StateKey` is now `StateKey(domain)` only.

Closes #717

### CODE REVIEW

**Comment by NodeJSmith** [comment]
@coderabbitai full review

File: src/hassette/core/state_proxy.py [hunk 1]
```diff
@@ -417,23 +402,6 @@ def iter_states() -> Generator[tuple[str, "HassStateDict"], Any, None]:
 
         return iter_states()
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/core/state_proxy.py:403]: Fixed — migrated the stray `in state_proxy` containment check to `state_proxy.get_state(ENTITY) is not None`. This was the root cause of all four failing `system` CI checks.

File: codegen/src/hassette_codegen/extractors/constants.py [hunk 2]
```diff
@@ -34,36 +54,126 @@ def extract_sensor_constants(ha_core_path: Path) -> list[ExtractedConstantSet]:
...
+    Shared by every extractor below: each reads an upstream Home Assistant source file that may
+    not parse (an unexpected format change, a partial checkout), and each must treat that as "no
+    members found" rather than propagate the exception.
+    """
     source = filepath.read_text(encoding="utf-8")
     try:
-        tree = ast.parse(source, filename=str(filepath))
+        return ast.parse(source, filename=str(filepath))
     except SyntaxError:
+        return None
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply codegen/src/hassette_codegen/extractors/constants.py]: Fixed — `_parse_module_or_none` and `extract_numeric_state_expected_source` in `constants.py` now catch `OSError` alongside `SyntaxError`; `_check_predicate_freshness` in `pipeline.py` now catches `OSError` on the snapshot read and returns `False` with a warning.

File: codegen/src/hassette_codegen/pipeline.py [hunk 3]
```diff
@@ -182,6 +182,18 @@ def run_pipeline(
...
+    if check_mode and sensor_init_py.exists():
+        # Not a generated-file comparison — this guards the *hand-written* port in
+        # sensor_shapes.py against HA changing the logic it was ported from. The fixture test
+        # that pins hassette's own behavior can't see that; this can. Scoped to --check (what CI
+        # runs) so a plain `generate` run is never blocked by unrelated upstream drift, and
+        # scoped to sources that actually have a sensor component so synthetic single-domain
+        # test fixtures elsewhere in this suite are unaffected.
+        numeric_predicate_snapshot = repo_root / "codegen" / "snapshots" / "numeric_state_expected.py.txt"
+        if not _check_predicate_freshness(ha_source.path, numeric_predicate_snapshot):
+            any_drift = True
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply codegen/src/hassette_codegen/pipeline.py]: Fixed — the guard now checks `sensor_dir.is_dir()` instead of `sensor_init_py.exists()`, so a relocated/renamed `__init__.py` is reported as drift by `_check_predicate_freshness()` instead of silently skipped.

File: codegen/tests/test_overrides.py [hunk 4]
```diff
@@ -3,9 +3,19 @@
...
-from hassette_codegen.overrides import DomainOverride, get_override, load_overrides, validate_overrides
+from hassette_codegen.extractors.properties import ExtractedProperty
+from hassette_codegen.overrides import (
+    DomainOverride,
+    PropertyOverride,
+    apply_property_overrides,
+    get_override,
+    load_overrides,
+    validate_overrides,
+)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply codegen/tests/test_overrides.py:18]: Declining for this PR — verified the specific mechanism (this repo's codegen has no pytest `pythonpath` config; `hassette-codegen` is installed editable in the workspace venv, which is why the import already works without `sys.path.insert`), but this identical pattern appears in all 14 codegen test files, predating this branch (this file only had test content added, not the import block). Fixing one file in isolation would be inconsistent with its 13 siblings — a repo-wide cleanup is out of scope here.

File: docs/pages/core-concepts/states/index.md [hunk 5]
```diff
@@ -118,11 +118,72 @@ Hassette auto-generates typed state classes for 55 Home Assistant domains from H
...
-- **`states.BinarySensorState`** has `value: bool | None`, `attributes.device_class: str | None`
+- **`states.SensorState`** has `value: str | None`, `attributes.unit_of_measurement: str | None`, `attributes.device_class: SensorDeviceClass | None`
+- **`states.BinarySensorState`** has `value: bool | None`, `attributes.device_class: BinarySensorDeviceClass | None`
 
 The API reference lists all 55 classes with their full attribute signatures. Domains not covered there are handled by [Custom States](custom-states.md).
 
+## Sensor Value Shapes
+
+`SensorState.value` is `str | None`, because the `sensor` domain covers every kind of sensor Home Assistant supports — a door lock's battery percentage and a smart meter's next-reset timestamp are both `sensor.*` entities. Home Assistant tells them apart with `device_class`, and that metadata sorts every sensor into one of four value shapes: numeric, enum, timestamp, or date. Four state classes narrow `value` to match:
+
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply docs/pages/core-concepts/states/index.md:129]: Fixed — applied the suggested wording, noting sensors without sufficient metadata remain `UNKNOWN`.

File: docs/pages/core-concepts/states/index.md [hunk 6]
```diff
@@ -118,11 +118,72 @@ Hassette auto-generates typed state classes for 55 Home Assistant domains from H
...
+`.get()` and `[]` diverge on a narrowed accessor the same way they do everywhere else in `DomainStates`, with one addition: a non-member entity is a distinct case from a missing one.
+
+```python
+--8<-- "pages/core-concepts/states/snippets/sensor_shapes.py:lookup"
+```
+
+`.get()` returns `None` for an entity that exists but isn't a shape-view member — `EntityNotInViewError` subclasses `KeyError`, and `Mapping.get()` is implemented by catching `KeyError`. Direct `[]` access raises `EntityNotInViewError` instead, naming the entity, its actual `device_class`, and the shape the accessor expected. Iteration skips non-members silently; a conversion failure — metadata matches the shape but the value fails to parse — is additionally logged at `debug`.
+
+!!! note "`self.states[NumericSensorState]` is unsupported"
+    The four shape classes deliberately don't declare their own `domain` — doing so would register them in place of `SensorState` process-wide. That means generic indexing (`self.states[NumericSensorState]`) raises `NoDomainAnnotationError` rather than working. Use the dedicated accessor (`self.states.numeric_sensor`) instead; the error message names it.
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply docs/pages/core-concepts/states/index.md]: Fixed — applied the suggested declarative wording.

File: src/hassette/conversion/state_registry.py [hunk 7]
```diff
@@ -245,6 +267,16 @@ def convert_state_dict_to_model(value: typing.Any, model: "type[BaseState]") ->
...
 
+    entity_id_str = str(entity_id) if entity_id else "<unknown>"
+    attributes = prepared.get("attributes") or {}
+    device_class = attributes.get("device_class")
+
+    expected_shape = NARROWED_SENSOR_SHAPES.get(model)
+    if expected_shape is not None:
+        actual_shape = classify_sensor_shape(attributes)
+        if actual_shape is not SensorShape.UNKNOWN and actual_shape is not expected_shape:
+            raise SensorShapeMismatchError(entity_id_str, device_class, model)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/conversion/state_registry.py:284]: Fixed — `convert_state_dict_to_model` now only classifies when `attributes` is a `Mapping`, letting `model_validate()` reject a malformed value with `UnableToConvertAnnotatedStateError` as documented; `classify_sensor_shape` now guards the `_KNOWN_DEVICE_CLASSES` membership check with `isinstance(device_class, str)` so an unhashable value (e.g. a list) no longer raises `TypeError`. Added regression tests for both.

File: src/hassette/models/states/catalog.py [hunk 8]
```diff
@@ -22,42 +22,28 @@ class StateKey:
...
+    """Register a state class for a specific domain.
 
     Args:
         state_class: The state class to register. Must be a subclass of BaseState.
         domain: The Home Assistant domain (e.g., "light", "sensor").
-        device_class: The device class (e.g., "temperature", "motion").
     """
-    key = StateKey(domain=domain, device_class=device_class)
+    key = StateKey(domain=domain)
     _STATE_CATALOG[key] = state_class
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/models/states/catalog.py:43]: Fixed — added `DomainRequiredError`, raised from `register_state_converter()` when `domain is None`. `StateRegistry.register()`'s `domain` param is now required-keyword instead of defaulting to `None`. Also found and fixed one existing test call site that was silently writing a `StateKey(domain=None)` entry into the catalog.

File: src/hassette/state_manager/state_manager.py [hunk 9]
```diff
@@ -26,6 +27,27 @@
...
+    """Build a membership predicate for one of the four narrowed sensor-shape accessors.
+
+    Each predicate re-runs :func:`classify_sensor_shape` per access rather than caching a
+    result, so a runtime ``device_class`` change is picked up on the very next access — see
+    ``design/specs/093-sensor-device-class-subtypes/design.md`` Edge Cases, "device_class
+    changes at runtime."
+    """
+
+    def predicate(state: "HassStateDict") -> bool:
+        return classify_sensor_shape(state.get("attributes", {})) == shape
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/state_manager/state_manager.py]: Fixed — both call sites now use `state.get("attributes") or {}` to guard against an explicit `"attributes": null`, matching the conversion path's existing hardening.

File: src/hassette/state_manager/state_manager.py [hunk 10]
```diff
@@ -110,43 +162,46 @@ def to_dict(self) -> dict[str, StateT]:
...
     def __len__(self) -> int:
-        """Return the number of entities in this domain."""
-        return self._state_proxy.num_domain_states(self._domain)
+        """Return the number of member entities in this domain.
+
+        Computed by iterating and checking membership (predicate AND convertibility) rather than
+        consulting the state proxy's raw domain count, so this always agrees with ``list(self)`` —
+        see ``_validate_if_member``.
+        """
+        return sum(1 for _ in self)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/state_manager/state_manager.py:177]: Fixed — `__bool__` now short-circuits with `next(iter(self), None) is not None` instead of calling `len(self)`.

File: tests/integration/test_sensor_shape_accessors.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        hassette = hassette_with_state_proxy
+        await seed_multi_shape_sensors(hassette)
+
+        states_instance = StateManager(hassette.hassette, parent=hassette.hassette)
+        date_sensors = states_instance.date_sensor
+
+        assert isinstance(date_sensors, DomainStates)
+        assert set(date_sensors) == {"sensor.expiry"}
+        for state in date_sensors.values():
+            assert isinstance(state, DateSensorState)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/integration/test_sensor_shape_accessors.py:116]: Declining for this PR — agreed this is a real duplication, but it's tagged "Low value" and is pure test-only construction with no behavior risk, so not worth the scope expansion here.

File: tests/pyright_probes/sensor_shape_probe/narrowing_probe.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+"""Pyright probe proving `NumericSensorState.value` narrows to `float | None`.
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/pyright_probes/sensor_shape_probe/narrowing_probe.py]: Fixed — corrected the docstring summary to say the narrowed type is `float`, matching the docstring body below it.

File: tests/unit/state_manager/test_domain_states_membership.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        # root logger, so without this the child logger's records never reach it. Same workaround
+        # as tests/unit/web/test_middleware.py and tests/unit/core/conftest.py.
+        logging.getLogger("hassette").propagate = True
+
+        with caplog.at_level(logging.DEBUG, logger="hassette.state_manager.state_manager"):
+            list(ds)
+
+        state_manager_records = [r for r in caplog.records if r.name == "hassette.state_manager.state_manager"]
+        assert all(r.levelno < logging.ERROR for r in state_manager_records)
+        assert any(r.levelno == logging.DEBUG and "sensor.garbage" in r.getMessage() for r in state_manager_records)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/unit/state_manager/test_domain_states_membership.py]: Fixed — applied the suggested refactor: removed both `caplog`-based tests (one asserted on log content, the other set `logging.getLogger("hassette").propagate = True` and never restored it, leaking into later tests), replaced with a single membership-only assertion, dropped the now-unused `import logging`, and renamed the class to `TestPredicateExclusions`. Also caught and fixed a related drift while in this file: a local test-only `numeric_shape_predicate` mirror still used the pre-fix `state.get("attributes", {})` form after the null-safety fix landed elsewhere in this PR — deleted it, every test now exercises the real production predicate.
